test/unit/mdc-toolbar/foundation.test.js

```diff
@@ -0,0 +1,676 @@
+
+  setDeviceDesktop(mockAdapter);
+  foundation.init();
+
+  assert.equal(foundation.calculations_.toolbarRowHeight, rowHeight);
+  assert.equal(foundation.calculations_.toolbarRatio, 5);
+  assert.equal(foundation.calculations_.flexibleExpansionRatio, 3);
+  assert.equal(foundation.calculations_.maxTranslateYRatio, 0);
+  assert.equal(foundation.calculations_.scrollThesholdRatio, 3);
+});
+
+test('#init sets correct ratio and style for fixed last row flexible header', () => {
+  const {foundation, mockAdapter} = setupTest();
+
+  const rowHeight = numbers.TOOLBAR_ROW_HEIGHT;
+  const flexibleRowMaxHeight = rowHeight * 4;
+  const toolbarHeight = flexibleRowMaxHeight + rowHeight;
+  td.when(mockAdapter.hasClass(cssClasses.FIXED_LASTROW)).thenReturn(true);
+  td.when(mockAdapter.hasClass(cssClasses.TOOLBAR_ROW_FLEXIBLE)).thenReturn(true);
+  td.when(mockAdapter.getFlexibleRowElementOffsetHeight()).thenReturn(flexibleRowMaxHeight);
+  td.when(mockAdapter.getOffsetHeight()).thenReturn(toolbarHeight);
+
+  setDeviceDesktop(mockAdapter);
+  foundation.init();
+
+  assert.equal(foundation.calculations_.toolbarRowHeight, rowHeight);
+  assert.equal(foundation.calculations_.toolbarRatio, 5);
+  assert.equal(foundation.calculations_.flexibleExpansionRatio, 3);
+  assert.equal(foundation.calculations_.maxTranslateYRatio, 1);
+  assert.equal(foundation.calculations_.scrollThesholdRatio, 4);
+});
+
+test('#updateAdjustElementStyles adjust marginTop for fixed toolbar', () => {
+  const {foundation, mockAdapter} = setupTest();
+
+  td.when(mockAdapter.hasClass(cssClasses.FIXED)).thenReturn(true);
+  foundation.init();
+  foundation.updateAdjustElementStyles();
+
+  td.verify(mockAdapter.setStyleForFixedAdjustElement('marginTop', td.matchers.anything()));
+});
+
+test('#updateAdjustElementStyles adjust marginTop for fixed last row only toolbar', () => {
+  const {foundation, mockAdapter} = setupTest();
+
+  td.when(mockAdapter.hasClass(cssClasses.FIXED_LASTROW)).thenReturn(true);
+  foundation.init();
+  foundation.updateAdjustElementStyles();
+
+  td.verify(mockAdapter.setStyleForFixedAdjustElement('marginTop', td.matchers.anything()));
+});
+
+test('#updateAdjustElementStyles do not adjust marginTop for non-fixed toolbar', () => {
+  const {foundation, mockAdapter} = setupTest();
+
+  foundation.init();
+  foundation.updateAdjustElementStyles();
+
+  td.verify(mockAdapter.setStyleForFixedAdjustElement('marginTop', td.matchers.anything()), {times: 0});
+});
+
+test('on scroll debounces calls within the same frame', () => {
+  const {foundation, mockAdapter} = setupTest();
+  const mockRaf = createMockRaf();
+  const {scrollHandler} = mockHandler(foundation, mockAdapter, mockRaf);
+
+  scrollHandler();
+  scrollHandler();
+  scrollHandler();
+  assert.equal(mockRaf.pendingFrames.length, 1);
+  mockRaf.restore();
+});
+
+test('on scroll resets debounce latch when updateToolbarStyles_ frame is run', () => {
```

Comment:
supernit: I would avoid even referencing private property names in test descriptions. Basically, a private property/method should only ever be known _within_ the component, and never referred to within the "outside world" of a component. This way if the private property/method name changes, this test description won't be out of date.